crates/composition/src/composition_ir.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    push_and_return_id,
+    subgraphs::{self, StringWalker},
+};
+use grafbase_federated_graph as federated;
+use std::collections::{BTreeSet, HashMap};
+
+/// The intermediate representation of the schema that is produced by composition. This data
+/// structure is distinct from [FederatedGraph](grafbase_federated_graph::FederatedGraph) because
+/// it is accumulated out of order during composition. Only after all composed top-level
+/// definitions (objects, interfaces, scalars, enums, unions) are known can we construct
+/// [federated::FieldTypeId]s and [federated::ObjectId]s for union members for the federated graph.
+///
+/// This is a **write only** data structure during composition. The source of truth for the
+/// contents of the federated graph is the subgraphs.
+#[derive(Default)]
+pub(crate) struct CompositionIr {
+    pub(crate) field_type_kinds_map: HashMap<subgraphs::StringId, federated::FieldTypeKind>,
```

Comment:
Todo: rename this, the name is bad. Also maybe `FieldTypeKind` should be renamed to `DefinitionId`?

Reply:
Agreed, I'm not really clear what this field is from it's name

Reply:
Particularly the `field` prefix - kind of looks like it's just a map from types to kinds?  

Though maybe I'm misunderstanding - as I said, finding it quite difficult to follow all the functions in here with the nested helper functions

Reply:
It's a map from a name to a field type (but field type was already taken, so it became field type kind, which isn't a good name). I'll at least add a comment here and maybe more in other places, and rename the field and the enum.

Reply:
Where field type is actually a definition, so a scalar, or an enum, or an interface, etc.